### PR TITLE
Summarize subcommands

### PR DESCRIPTION
**Pull request by anntzer**
For a program with multiple subcommands, such as
```
import defopt

def foo():
    """
    A summary.

    More help.
    """

def bar():
    """
    Another summary.

    And more help.
    """

defopt.run(foo, bar)
```
this patch improves the help displayed to
```
$ python test.py -h
usage: test.py [-h] {foo,bar} ...

positional arguments:
  {foo,bar}
    foo       A summary.
    bar       Another summary.

optional arguments:
  -h, --help  show this help message and exit

$ python test.py foo -h
usage: test.py foo [-h]

A summary.

More help.

optional arguments:
  -h, --help  show this help message and exit
```

Also split out the parser-and-main creation part in a separate function,
and change the signature of `_get_doc` to take the docstring as
argument, to allow caching of the parse (1. it is technically possible
to use the functions as keys in the cache, but one could change
the docstrings of the functions; 2. unfortunately, one cannot do
`parser.help = _get_doc_summary(doc.text)` in `_populate_parser` --
setting the subparser help is more intricate).

Depends on #23 -- only the second commit is new.

### CODE REVIEW

**Comment by evanunderscore** [comment]
Great idea - I hadn't realized this was missing. I'll find some time to review this soon.

**Comment by anntzer** [comment]
Rebased following the merge of the automatic short flag PR.

**Comment by anntzer** [comment]
bump :)

**Comment by evanunderscore** [comment]
This is still on my list of things to do, I just don't have a lot of spare time at the moment. Thanks for your patience.

**Comment by anntzer** [comment]
No worries.

**Comment by evanunderscore** [comment]
Your reasoning for changing the signature of `_parse_doc` is sound, but it does make the code a little messier and adds a lot of noise to the diff. Could you instead change `_parse_doc` to do the `inspect.getdoc` call then pass that through to a new helper with most of the current code?

`_create_parser_and_main` has ended up being a little bit awkward because of how the code was initially. I would prefer it if we could make the `main` part go away by setting the `_func` default in the single function case too, that way there's no need to return a function or branch based on whether it is `None`. I don't mind if you don't feel like changing this right now though.

`_get_doc_summary` feels backwards to me - we're building `text` by joining the paragraphs together only to split them back out again. Since the `_Doc` namedtuple is entirely private, I think it makes more sense to add another member to that for the summary and set it inside `_parse_doc` before they're joined.

Overall great work, let me know if you disagree with anything I've said and we can discuss it.

**Comment by anntzer** [comment]
Reasonable and easy :-) (except the second point, which I'll leave for now)
I subtly updated the commit message too.

File: test_defopt.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from argparse import ArgumentParser
 from enum import Enum
+import inspect
```
Comment (evanunderscore) [inline test_defopt.py]: This import is no longer necessary.

Comment (anntzer) [inline test_defopt.py]: fixed

**Comment by evanunderscore** [comment]
This looks great, thanks! Is there anything else you'd like to change before I push up a new release?

**Comment by anntzer** [comment]
Yes.

I would like to change the signature of `defopt.run` so that instead of taking `*funcs` and choosing to dispatch on subcommands if there is more than one func, it instead takes exactly one positional-only argument, which is either a single function or a list of functions, and dispatches on subcommands if a list is passed.

This would allow dispatching even with only a single subcommand, which can be useful e.g. if more subcommands are planned and simply not implemented yet.  (I also think that behavior that depends on the number of arguments passed in `*args` -- as opposed to the type of a specific argument -- is relatively less pythonic.)

Of course this is backwards incompatible, so if you really want you can keep both behaviors available, so that:
`defopt.run(func)` -> no subcommand
`defopt.run(func1, func2, ...)` -> subcommands
`defopt.run([func])` -> subcommand
`defopt.run([func1, func2])` -> subcommands

Sounds reasonable?

**Comment by evanunderscore** [comment]
Sounds good to me. I've had the same thought before but never had a pressing need to implement it. I would like to go with the backwards-compatible option, and emit a `DeprecationWarning` when more than one argument is passed positionally.

**Pull request merged**